zcash_primitives/src/transparent.rs

```diff
@@ -0,0 +1,105 @@
+use crate::{legacy::TransparentAddress, sapling::keys::prf_expand_vec};
+use hdwallet::{traits::Deserialize, ExtendedPrivKey, ExtendedPubKey};
+use sha2::{Digest, Sha256};
+use std::convert::TryInto;
+
+/// A type representing a private key at the BIP-44 external child
+/// level `m/44'/<coin_type>'/<account>'/0/<child_index>
+#[derive(Clone, Debug)]
+pub struct ExternalPrivKey(pub ExtendedPrivKey);
+
+impl ExternalPrivKey {
+    /// Returns the external public key corresponding to this private key
+    pub fn to_external_pubkey(&self) -> ExternalPubKey {
+        ExternalPubKey(ExtendedPubKey::from_private_key(&self.0))
+    }
+
+    /// Extracts the secp256k1 secret key component
+    pub fn secret_key(&self) -> &secp256k1::key::SecretKey {
+        &self.0.private_key
+    }
+}
+
+pub fn pubkey_to_address(pubkey: &secp256k1::key::PublicKey) -> TransparentAddress {
+    let mut hash160 = ripemd160::Ripemd160::new();
+    hash160.update(Sha256::digest(&pubkey.serialize()));
+    TransparentAddress::PublicKey(*hash160.finalize().as_ref())
+}
+
+/// A type representing a public key at the BIP-44 external child
+/// level `m/44'/<coin_type>'/<account>'/0/<child_index>
+#[derive(Clone, Debug)]
+pub struct ExternalPubKey(pub ExtendedPubKey);
+
+impl std::convert::TryFrom<&[u8]> for ExternalPubKey {
+    type Error = hdwallet::error::Error;
+
+    fn try_from(data: &[u8]) -> Result<Self, Self::Error> {
+        let ext_pub_key = ExtendedPubKey::deserialize(data)?;
+        Ok(Self(ext_pub_key))
+    }
+}
+
+impl ExternalPubKey {
+    /// Returns the transparent address corresponding to
+    /// this public key.
+    pub fn to_address(&self) -> TransparentAddress {
+        pubkey_to_address(&self.0.public_key)
+    }
+
+    /// Returns the secp256k1::key::PublicKey component of
+    /// this public key.
+    pub fn public_key(&self) -> &secp256k1::key::PublicKey {
+        &self.0.public_key
+    }
+
+    /// Returns the chain code component of this public key.
+    pub fn chain_code(&self) -> &[u8] {
+        &self.0.chain_code
+    }
+
+    /// Derives the internal ovk and external ovk corresponding to this
+    /// transparent fvk. As specified in [ZIP 316][transparent-ovk].
+    ///
+    /// [transparent-ovk]: https://zips.z.cash/zip-0316#deriving-internal-keys
+    fn ovk_for_shielding(&self) -> (InternalOvk, ExternalOvk) {
```

Comment:
This is a bug in the PR: `ExternalPubKey` is derived at the address level, but ZIP 316 specifies that the internal and external OVKs must be derived from the transparent FVK, which is at the account level.